### ISSUE TITLE
Hook version of pre-commit should be compared against current version of pre-commit

### ISSUE THREAD

**Issue by arkban** [pre-commit/pre-commit#1305]
TL;DR The version of pre-commit referenced in the hook should be compared against current version of pre-commit.

I've been using pre-commit for a while, and I didn't realize that the git hook hard codes the version of pre-commit in the hook, which may not equal the currently installed version of pre-commit, e.g.:
```shell
$ pre-commit --version
pre-commit 2.0.1

$ rg pre-commit .git
...
.git/hooks/pre-commit
...
19:INSTALL_PYTHON = '/usr/local/Cellar/pre-commit/1.20.0/libexec/bin/python3.7'
...
```
_Note: `rg` is [ripgrep](https://github.com/BurntSushi/ripgrep) which is just fancy/pretty {{grep}}_ 

The version in the git hook is **1.20.0** where the version I had installed is **2.0.1**. 

I followed this rabbit hole by doing this first:
```
$ pre-commit run -a
[WARNING] Unexpected key(s) present on git://github.com/pre-commit/pre-commit-hooks: sha
[WARNING] Unexpected key(s) present on git@github.com:AudaxHealthInc/pre-commit-hooks.git: sha
```
I dutifully changed my `sha` to `rev`, tried to commit, and got an errors like this:
```shell
$ git commit --amend --no-edit
An error has occurred: InvalidConfigError:
==> File /Users/.../.pre-commit-config.yaml
==> At Repository(repo='git://github.com/pre-commit/pre-commit-hooks')
=====> Missing required key: sha
Check the log at ~/.pre-commit/pre-commit.log
```

So it would be really nice if there was some kind of version check, or even if the git hook could use `which pre-commit` or something, to avoid version mismatches.

**Comment by arkban** [comment]
Oh, I forgot to mention the fix for this is simple:
```
$ pre-commit uninstall
pre-commit uninstalled

$ pre-commit install
pre-commit installed at .git/hooks/pre-commit
```

And I verified the version is correct now:
```
$ rg pre-commit .git
...
20:INSTALL_PYTHON = '/usr/local/Cellar/pre-commit/2.0.1/libexec/bin/python3.8'
....
```

**Comment by asottile** [comment]
the script in `.git/hooks/pre-commit` doesn't matter -- it doesn't change what actual version is invoked.

the `sha` key has been warning since [`v1.17.0`](https://github.com/pre-commit/pre-commit/blob/master/CHANGELOG.md#1170---2019-06-06) and deprecated since [v1.7.0](https://github.com/pre-commit/pre-commit/blob/master/CHANGELOG.md#170---2018-03-03)

the `INSTALL_PYTHON` is just one of the ways that the `.git/hooks/pre-commit` script attempts to bootstrap pre-commit -- if that exists it'll use it, otherwise it'll look for a `pre-commit` executable.  If `INSTALL_PYTHON` is out of date it isn't the end of the world and doesn't really affect anything

In your case you have two copies of pre-commit:
1. some obscenely old version installed to `/usr/bin/python` (this one is no longer possible to be chosen as of pre-commit 2.x)
2. the version installed by brew

the old (<2.x) approach used to try three things to discover pre-commit:
1. `INSTALL_PYTHON`
2. `sys.executable`
3. `which pre-commit`

2. is no longer attempted due to the unlikelihood of it succeeding (however, looks like on macos you've got quite the mess!)

you can think of the script in `.git/hooks/pre-commit` as a bit of a "bootstrap script" -- it doesn't actually have access to any of the pre-commit code (including the version number!!!) and is mostly looking for a way to find it and then invoke it.  the version number you're seeing is because `brew` installs things to versioned directories

**Comment by arkban** [comment]
> In your case you have two copies of pre-commit:
> 1. some _obscenely_ old version installed to ...

I will take that as a compliment ; ) and claim it is a testament to `pre-commit`'s excellence.

> the version number you're seeing is because brew installs things to versioned directories

Ahh, so the version being in the hook script is an artifact of the fact that I'm using `brew`. Okay, that makes sense.

**Comment by asottile** [comment]
indeedy!
